### ISSUE TITLE
ETL of PubMed article titles into KG2

### ISSUE THREAD

**Issue by saramsey** [RTXteam/RTX#646]
For each PMID that is referenced anywhere in KG2, include a node for the PMID with the article title as the `name` field. Requested by the UAB/PMI team.

**Comment by ericawood** [comment]
@saramsey I have created a script that pulls out the PMIDs referenced in KG2 and saves them for my other script that parses the PubMed XML files. It is about 303M large. I also wrote a script that downloads the PubMed XML dump from the ftp site, but it is very large (it was about 144G before I stopped it for fear of running out of disk space on kg2erica). Does pubmed.rtx.ai already have the dump?

**Comment by edeutsch** [comment]
I believe the answer is yes, but I do not know details. @amykglen would.

**Comment by amykglen** [comment]
yep - the most recent pubmed xml dump is in `/home/ubuntu/ultrafast-ngd-amy/pubmed_xml_files_2020` on `pubmed.rtx.ai` (consists of about 1,000 `xml.gz` files).

though it looks like it's only 28G in total, so I'm not sure why yours was so large, @ericawood. (I used `wget -r ftp://ftp.ncbi.nlm.nih.gov/pubmed/baseline` to download it, which took a little less than an hour.)

**Comment by ericawood** [comment]
@saramsey What is the maximum KG JSON file size we are looking for with the PubMed article titles and corresponding MESH edges?
Even without the journal and author names in the graph, my test KG with PMID nodes for every PMID currently in KG2 and their MESH edges was 76GB large.

**Comment by saramsey** [comment]
> @saramsey What is the maximum KG JSON file size we are looking for with the PubMed article titles and corresponding MESH edges?
> Even without the journal and author names in the graph, my test KG with PMID nodes for every PMID currently in KG2 and their MESH edges was 76GB large.

Hi @ericawood thanks for the update. 

Hmm, that is big. A few things to check (I realize you may have already checked these points, but just in case).
(in my comments below, I'll call your 76 GiB JSON file `bigfile.json` for the sake of specificity):

1. is `bigfile.json` pretty-printed JSON or single-line ("compressed") JSON?
2. How many unique PMIDs are there that are referenced in KG2? (this will help us to interpret whether the JSON filesize estimate of the PMID-to-MESH mappings is likely to come down at all).
3. Does `bigfile.json` contain *only* edges?
4. Are we certain that only KG2-referenced PMIDs are in `bigfile.json`?
5. Are we certain that `bigfile.json` does not contain any duplicate edges?

**Comment by saramsey** [comment]
Ah, it seems like the 76 GiB filesize for `bigfile.json` includes both nodes and edges. If it also includes nodes for the PMIDs, how much of that is due to the abstracts themselves? We could certainly remove the abstracts since those are retrievable from PubMed on-demand (and probably less usable for reasoning within ARAX) anyhow.

**Comment by ericawood** [comment]
Hi Steve,
Thank you for your comments.

> is `bigfile.json` pretty-printed JSON or single-line ("compressed") JSON?

It is single-line ("compressed") JSON. I used `cat` to see a portion of it into a file and it looked like this:
![image](https://user-images.githubusercontent.com/36611732/88726910-66425c00-d0e3-11ea-975e-a5efcc7045d7.png)

> How many unique PMIDs are there that are referenced in KG2? (this will help us to interpret whether the JSON filesize estimate of the PMID-to-MESH mappings is likely to come down at all).

I wrote a script that extracted all of the PMIDs out of KG2 (the old version). Its output file contains 19014220 PMIDs.

> Does `bigfile.json` contain only edges?

`bigfile.json` contains both nodes (PMIDs) and edges (PMID->MESH)

> Are we certain that only KG2-referenced PMIDs are in bigfile.json?

When I tested my ETL script on a smaller subset of PubMed, it only contained PMIDs in KG2. (`bigfile.json` pulled from both [baseline](ftp://ftp.ncbi.nlm.nih.gov/pubmed/baseline) and [updateFiles](ftp://ftp.ncbi.nlm.nih.gov/pubmed/updatefiles)).

> Are we certain that bigfile.json does not contain any duplicate edges?

I will test this tomorrow, thank you for the suggestion!

**Comment by ericawood** [comment]
One problem with testing `bigfile.json` - it is so big that it can't be loaded into memory. (I tested this using the report_stats_on_json_kg.py script. It didn't get past loading the file in, so it presumably didn't finish). It got to using 99% of the memory, before going into swap, then hitting the error `Killed`. @saramsey Do you have any suggestions?

**Comment by ericawood** [comment]
Hi @saramsey , I reran my ETL script for the PMIDs and PMID->MESH edges. As it was saving the JSON file, it got a `Killed` error. (I had added in code to ensure the there would be no duplicate edges by JSON dumping them into strings to use as dict keys then JSON loading them back before it saved the file). The way I see it, we have three options going forward.
1. Don't bring any data from PubMed into KG2.
2. Only bring PMIDs (not PMID->MESH edges) into KG2.
3. Put an arbitrary limit on the number of MESH edges based on the goal output filesize.

Please let me know what you think.

**Comment by saramsey** [comment]
Hi @ericawood does `bigfile.json` also contain the *abstracts* of the pubmed articles?

**Comment by ericawood** [comment]
Hi @saramsey , `bigfile.json` does not contain the _abstracts_ of the pubmed articles.

**Comment by saramsey** [comment]
I'm inclined to create a half-terabyte EC2 instance to facilitate further development work on this. If I do that, can you rsync the data into it and so forth?

**Comment by saramsey** [comment]
Hi @ericawood just confirming, you are unable to load `bigfile.json` on a `r5.8xlarge` instance, correct?

**Comment by ericawood** [comment]
@saramsey What is rsync? Is that like scp? To get the data onto kg2erica, I used an extraction script that I wrote to download all of it (it also downloads the daily updates). Yes, I am unable to load `bigfile.json` on a `r5.8xlarge` and, this last time, I was also unable to generate it (I had added code to try to ensure there were no duplicate edges that ended up taking more memory).

**Comment by saramsey** [comment]
Hi @ericawood, `rsync` is a file synchronization utility that can accomplish the same thing as marking a tarball, scping the tarball to a new instance, and untarring the tarball in the new instance.

**Comment by ericawood** [comment]
Hi @saramsey, that sounds good. I will plan on using `rsync`. Thank you for your explanation!

**Comment by saramsey** [comment]
OK, for issue #646 (pubmed), I have created a `r5a.16xlarge` instance in `us-west-2` (see below). I've sent you the Manager Passcode by email. I have installed your RSA public key into it. It is currently stopped. It is a clean Ubuntu 18.04 instance.  Only port 22 is currently open but I can open any ports you need. Hostname is `bigmem.rtx.ai`.

<img width="725" alt="Screen Shot 2020-08-11 at 11 47 52 AM" src="https://user-images.githubusercontent.com/5562107/89936568-b849ae00-dbc8-11ea-8eaf-8501cd04c0d3.png">

**Comment by saramsey** [comment]
Your EC2 instance was archived to an Amazon Machine Image (AMI). Just let me know if you want to provision the instance again and it will be exactly as it was before (I hope).
